File: packages/sdk/hush-py/src/clawdstrike/hunt/local.py
"""Local offline envelope loading from filesystem directories.

Port of ``hunt-query/src/local.rs``.
"""

from __future__ import annotations

import json
import logging
from datetime import datetime, timezone
from pathlib import Path

from clawdstrike.hunt.duration import parse_human_duration
from clawdstrike.hunt.query import matches_query
from clawdstrike.hunt.timeline import merge_timeline, parse_envelope
from clawdstrike.hunt.types import (
    EventSourceType,
    HuntQuery,
    NormalizedVerdict,
    TimelineEvent,
)

logger = logging.getLogger(__name__)


def _normalize_utc(dt: datetime) -> datetime:
    """Normalize datetimes to timezone-aware UTC for safe comparisons."""
    if dt.tzinfo is None:
        return dt.replace(tzinfo=timezone.utc)
    return dt.astimezone(timezone.utc)


def default_local_dirs() -> list[str]:
    """Return default directories to search for local envelopes.

    Only directories that actually exist are included.
    """
    home = Path.home()
    candidates = [
        home / ".clawdstrike" / "receipts",
        home / ".clawdstrike" / "scans",
        home / ".hush" / "receipts",
    ]
    return [str(d) for d in candidates if d.is_dir()]


def query_local_files(
    query: HuntQuery,
    search_dirs: list[str] | None = None,
) -> list[TimelineEvent]:
    """Query envelopes from local JSON/JSONL files.

    Reads all ``.json`` and ``.jsonl`` files from the given directories,
    parses envelopes, filters with the query predicates, merges by timestamp,
    and truncates to the newest ``query.limit`` events.

    If *search_dirs* is ``None``, defaults to :func:`default_local_dirs`.
    """
    dirs = search_dirs if search_dirs is not None else default_local_dirs()
    all_events: list[TimelineEvent] = []

    for dir_str in dirs:
        dir_path = Path(dir_str)
        if not dir_path.is_dir():
            logger.debug("skipping non-directory: %s", dir_path)
            continue

        try:
            entries = list(dir_path.iterdir())
        except OSError as exc:
            logger.warning("skipping unreadable directory %s: %s", dir_path, exc)
            continue

        for entry in entries:
            if not entry.is_file():
                continue

            suffix = entry.suffix.lower()
            if suffix == ".jsonl":
                try:
                    events = _read_jsonl_file(entry)
                except (OSError, json.JSONDecodeError, ValueError, UnicodeDecodeError):
                    logger.warning("skipping unreadable/invalid JSONL file %s", entry)
                    continue
            elif suffix == ".json":
                try:
                    events = _read_json_file(entry)
                except (OSError, json.JSONDecodeError, ValueError, UnicodeDecodeError):
                    logger.warning("skipping unreadable/invalid JSON file %s", entry)
                    continue
            else:
                continue

            for event in events:
                if matches_query(query, event):
                    all_events.append(event)

    merged = merge_timeline(all_events)
    _truncate_to_newest(merged, query.limit)
    return merged


def hunt(
    *,
    sources: tuple[EventSourceType, ...] = (),
    verdict: NormalizedVerdict | None = None,
    start: datetime | str | None = None,
    end: datetime | None = None,
    action_type: str | None = None,
    process: str | None = None,
    namespace: str | None = None,
    pod: str | None = None,
    entity: str | None = None,
    limit: int = 100,
    dirs: list[str] | None = None,
) -> list[TimelineEvent]:
    """High-level convenience function for querying local events.

    Parses duration strings for *start* (e.g. ``"1h"``, ``"30m"``),
    applies defaults, builds a query, and calls :func:`query_local_files`.
    """
    start_dt: datetime | None = None
    if isinstance(start, str):
        td = parse_human_duration(start)
        if td is not None:
            start_dt = datetime.now(tz=timezone.utc) - td
    elif isinstance(start, datetime):
        start_dt = _normalize_utc(start)

    end_dt = _normalize_utc(end) if isinstance(end, datetime) else None

    query = HuntQuery(
        sources=sources,
        verdict=verdict,
        start=start_dt,
        end=end_dt,
        action_type=action_type,
        process=process,
        namespace=namespace,
        pod=pod,
        entity=entity,
        limit=limit,
    )

    return query_local_files(query, dirs)


def _truncate_to_newest(events: list[TimelineEvent], limit: int) -> None:
    """Truncate *in-place* keeping only the newest ``limit`` events."""
    if limit == 0:
        events.clear()
        return
    if len(events) > limit:
        del events[: len(events) - limit]


def _read_json_file(path: Path) -> list[TimelineEvent]:
    """Read a single JSON file as envelope(s)."""
    content = path.read_text(encoding="utf-8")
    value = json.loads(content)

    if isinstance(value, list):
        results: list[TimelineEvent] = []
        for item in value:
            if isinstance(item, dict):
                event = parse_envelope(item)
                if event is not None:
                    results.append(event)
        return results
    elif isinstance(value, dict):
        event = parse_envelope(value)
        return [event] if event is not None else []
    else:
        return []


def _read_jsonl_file(path: Path) -> list[TimelineEvent]:
    """Read a JSONL file (one JSON object per line)."""
    content = path.read_text(encoding="utf-8")
    events: list[TimelineEvent] = []

    for line in content.splitlines():
        trimmed = line.strip()
        if not trimmed:
            continue
        try:
            value = json.loads(trimmed)
        except (json.JSONDecodeError, ValueError):
            continue
        if isinstance(value, dict):
            event = parse_envelope(value)
            if event is not None:
                events.append(event)

    return events

File: packages/sdk/clawdstrike-hunt/src/local.ts
import { readdir, readFile, stat } from 'node:fs/promises';
import { homedir } from 'node:os';
import { join, extname } from 'node:path';

import type { HuntQuery, NormalizedVerdict, TimelineEvent, EventSourceType } from './types.js';
import { matchesQuery } from './query.js';
import { mergeTimeline, parseEnvelope } from './timeline.js';
import { parseHumanDuration } from './duration.js';

/**
 * Default directories to search for local envelopes.
 * Returns only directories that actually exist.
 */
export async function defaultLocalDirs(): Promise<string[]> {
  const home = homedir();
  const candidates = [
    join(home, '.clawdstrike', 'receipts'),
    join(home, '.clawdstrike', 'scans'),
    join(home, '.hush', 'receipts'),
  ];

  const result: string[] = [];
  for (const dir of candidates) {
    try {
      const s = await stat(dir);
      if (s.isDirectory()) {
        result.push(dir);
      }
    } catch {
      // Directory does not exist, skip
    }
  }
  return result;
}

function truncateToNewest(
  events: TimelineEvent[],
  limit: number,
): TimelineEvent[] {
  if (limit === 0) {
    return [];
  }
  if (events.length <= limit) {
    return events;
  }
  return events.slice(events.length - limit);
}

/**
 * Query envelopes from local JSON/JSONL files.
 *
 * Reads files from the given search directories. `.json` files may contain
 * a single envelope or an array. `.jsonl` files contain one envelope per line.
 * Corrupt files/lines are skipped.
 * When `verify` is true, each parsed envelope gets `signatureValid` populated.
 *
 * Results are filtered by the query, merged by timestamp, and truncated
 * to the newest `query.limit` events.
 */
export async function queryLocalFiles(
  query: HuntQuery,
  searchDirs?: string[],
  verify: boolean = false,
): Promise<TimelineEvent[]> {
  const dirs = searchDirs ?? await defaultLocalDirs();
  const allEvents: TimelineEvent[] = [];

  for (const dir of dirs) {
    let isDir: boolean;
    try {
      const s = await stat(dir);
      isDir = s.isDirectory();
    } catch {
      isDir = false;
    }
    if (!isDir) {
      continue;
    }

    let entries: string[];
    try {
      entries = await readdir(dir);
    } catch {
      continue;
    }

    for (const entry of entries) {
      const filePath = join(dir, entry);
      let fileStat;
      try {
        fileStat = await stat(filePath);
      } catch {
        continue;
      }

      if (!fileStat.isFile()) {
        continue;
      }

      const ext = extname(entry).toLowerCase();
      let events: TimelineEvent[];

      if (ext === '.json') {
        try {
          events = await readJsonFile(filePath, verify);
        } catch {
          continue;
        }
      } else if (ext === '.jsonl') {
        try {
          events = await readJsonlFile(filePath, verify);
        } catch {
          continue;
        }
      } else {
        continue;
      }

      for (const event of events) {
        if (matchesQuery(query, event)) {
          allEvents.push(event);
        }
      }
    }
  }

  const merged = mergeTimeline(allEvents);
  return truncateToNewest(merged, query.limit);
}

/**
 * Options for the high-level `hunt()` convenience function.
 */
export interface HuntOptions {
  sources?: EventSourceType[];
  verdict?: NormalizedVerdict;
  /** Accepts a Date or a human-duration string like "1h", "30m". */
  start?: Date | string;
  end?: Date;
  actionType?: string;
  process?: string;
  namespace?: string;
  pod?: string;
  entity?: string;
  limit?: number;
  dirs?: string[];
  /** Verify envelope signatures and populate `event.signatureValid`. */
  verify?: boolean;
}

/**
 * High-level convenience function for querying local events.
 *
 * Parses duration strings for `start`, applies defaults, builds a query,
 * and calls `queryLocalFiles`.
 */
export async function hunt(options?: HuntOptions): Promise<TimelineEvent[]> {
  const opts = options ?? {};

  let startDate: Date | undefined;
  if (typeof opts.start === 'string') {
    const ms = parseHumanDuration(opts.start);
    if (ms !== undefined) {
      startDate = new Date(Date.now() - ms);
    }
  } else {
    startDate = opts.start;
  }

  const query: HuntQuery = {
    sources: opts.sources ?? [],
    verdict: opts.verdict,
    start: startDate,
    end: opts.end,
    actionType: opts.actionType,
    process: opts.process,
    namespace: opts.namespace,
    pod: opts.pod,
    entity: opts.entity,
    limit: opts.limit ?? 100,
  };

  return queryLocalFiles(query, opts.dirs, opts.verify ?? false);
}

async function readJsonFile(path: string, verify: boolean): Promise<TimelineEvent[]> {
  const content = await readFile(path, 'utf-8');
  const value: unknown = JSON.parse(content);

  if (Array.isArray(value)) {
    return value
      .map((v) => parseEnvelope(v, verify))
      .filter((e): e is TimelineEvent => e !== undefined);
  }

  const event = parseEnvelope(value, verify);
  return event !== undefined ? [event] : [];
}

async function readJsonlFile(path: string, verify: boolean): Promise<TimelineEvent[]> {
  const content = await readFile(path, 'utf-8');
  const events: TimelineEvent[] = [];

  for (const line of content.split('\n')) {
    const trimmed = line.trim();
    if (trimmed.length === 0) {
      continue;
    }
    try {
      const value: unknown = JSON.parse(trimmed);
      const event = parseEnvelope(value, verify);
      if (event !== undefined) {
        events.push(event);
      }
    } catch {
      // Skip invalid JSON lines
    }
  }

  return events;
}
